uservice_ccutter/server.py

```diff
@@ -1,61 +1,281 @@
+                                       userdict["github_repo"] + " failed.")
+                retdict = {"repo_url": remote_url}
+                return retdict
+
+        def create_github_repository(auth, userdict):
+            """Create new repo at GH."""
+            ghub = github3.login(auth["username"], token=auth["password"])
+            try:
+                ghub.me()
+            except github3.exceptions.AuthenticationFailed:
+                raise BackendError(status_code=401,
+                                   reason="Bad credentials",
+                                   content="Github login failed.")
+            namespc = userdict["github_repo"]
+            org, name = namespc.split('/')
+            desc = ""
+            if "description" in userdict:
+                desc = userdict["description"]
+            gh_org = None
+            for gorg in ghub.organizations():
+                if gorg.login == org:
+                    gh_org = gorg
+                    break
+            if gh_org is None:
+                raise BackendError(status_code=500,
+                                   reason="Internal Server Error",
+                                   content=auth["username"] +
+                                   " not in org " + org)
+            repo = gh_org.create_repository(name, description=desc)
+            if repo is None:
+                raise BackendError(status_code=500,
+                                   reason="Internal Server Error",
+                                   content="Github repository not created")
+            return repo.clone_url
+
+        def get_single_project_type(ptype):
+            """Return a single project type's cookiecutter.json."""
+            if ptype not in app.config["PROJECTTYPE"]:
+                types = [x for x in app.config["PROJECTTYPE"]]
+                raise BackendError(status_code=400,
+                                   reason="Bad Request",
+                                   content="Project type must be one of " +
+                                   str(types))
+            return app.config["PROJECTTYPE"][ptype]["template"]
+
+        def check_authorization():
+            """Sets app.auth["data"] if credentials provided, raises an
+            error otherwise."""
+            iauth = request.authorization
+            if iauth is None:
+                raise BackendError(reason="Unauthorized",
+                                   status_code=401,
+                                   content="No authorization provided.")
+            app.config["AUTH"]["data"]["username"] = iauth.username
+            app.config["AUTH"]["data"]["password"] = iauth.password
+
+        if run_standalone:
+            app.run(host='0.0.0.0', threaded=True)
 
 
 def standalone():
     """Entry point for running as its own executable."""
     server(run_standalone=True)
 
 
+def _refresh_cache(app, temp_dir, timeout):
+    """Refresh cookiecutter.json cache if needed."""
```

Comment:
I feel like a complete docstring is warranted here. Or at least document the caching behavior somewhere. There's a lot of logic in this function that's pretty hard to follow.